migrator-tool/src/main/scala/za/co/absa/spline/migrator/DataLineageToExecPlanWithEventConverter.scala

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.migrator
+
+import java.util.UUID
+
+import za.co.absa.spline.common.OptionImplicits._
+import za.co.absa.spline.migrator.DataLineageToExecPlanWithEventConverter.{AppMetaInfo, ExecutionEventExtra, ExecutionPlanExtra, OperationParams}
+import za.co.absa.spline.model.{DataLineage, op}
+import za.co.absa.spline.producer.rest.model._
+
+class DataLineageToExecPlanWithEventConverter(lineage: DataLineage) {
+
+  private val operationIdByDatasetUUID: Map[UUID, Int] =
+    lineage
+      .datasets
+      .map(_.id)
+      .zipWithIndex
+      .toMap
+
+  def convert(): (ExecutionPlan, Option[ExecutionEvent]) = {
+    val executionId = lineage.rootDataset.id
+
+    val (opWrite, opReads, opOther) = splitOperationsByType(lineage.operations)
+
+    val operations = Operations(
+      write = convertWriteOperation(opWrite),
+      reads = opReads.map(convertReadOperation),
+      other = opOther.map(convertOtherOperation))
+
+    val executionPlan = ExecutionPlan(
+      id = executionId,
+      operations = operations,
+      systemInfo = SystemInfo(AppMetaInfo.Spark, lineage.sparkVer),
+      agentInfo = Some(AgentInfo(AppMetaInfo.Spline, AppMetaInfo.SplineVersion)),
+      extraInfo = Map(
+        ExecutionPlanExtra.AppName -> lineage.appName,
+        ExecutionPlanExtra.DataTypes -> lineage.dataTypes,
+        ExecutionPlanExtra.Attributes -> lineage.attributes,
+        ExecutionPlanExtra.Datasets -> lineage.datasets
+      ))
+
+    val maybeExecutionEvent =
+      if (lineage.writeIgnored) None
+      else Some(
+        ExecutionEvent(
+          planId = executionId,
+          timestamp = lineage.timestamp,
+          error = None,
+          extra = Map(
+            ExecutionEventExtra.AppId -> lineage.appId,
+            ExecutionEventExtra.WriteMetrics -> opWrite.writeMetrics,
+            ExecutionEventExtra.ReadMetrics -> opWrite.readMetrics
+          ).filterNot(entryValueEmptyMapPredicate)))
+
+    (executionPlan, maybeExecutionEvent)
+  }
+
+  private val entryValueEmptyMapPredicate: ((_, _)) => Boolean = {
+    case (_, v: Map[_, _]) => v.isEmpty
+    case _ => false
+  }
+
+  private def splitOperationsByType(operations: Seq[op.Operation]): (op.Write, Seq[op.Read], Seq[op.Operation]) = {
+    assume(operations.nonEmpty)
+    val opWrite = operations.head.asInstanceOf[op.Write]
```

Comment:
Well, yes, and I totally agree with you. That's partially why I put `assume` there - to make it stinking just a little less.
I have considered several options before doing it. `isInstanceOf` is inevitable as there is no method on the `DataLineage` class that returns a `Write` type. I could add another case-clause to the following `foldLeft`, but I would end up with a collection of write operation with de-facto the only element which I would have to retrieve by `head` anyway. I could also use `collectFirst` with a partial function defined on `Write` and get an `Option[Write]`. But again, I would have to either use `Option.get` or I'd have to pull that option through the entire migrator logic to the place where the conversion result is actually used.
It all smells one way or another, so I thought that this way would be the least evil. Basically, we are dealing with the existing data model that implies certain conventions, and the head being a `Write` is one of them. So that one-liner says it all.